File: examples/kotlin/sse-client/src/main/kotlin/cdcexamples/sse/SseStream.kt
package cdcexamples.sse

/**
 * Event trägt ein Frame des Change-Streams (`LH-FA-SST-008`): den Namen aus
 * der `event:`-Zeile und die Nutzlast aus der `data:`-Zeile. Der Server
 * schreibt je Change ein JSON-Objekt mit den zehn Nachrichtenfeldern in die
 * `data:`-Zeile (`ADR-0061`).
 */
data class Event(val name: String, val data: String)

/**
 * SseStream baut die Stream-Adresse und zerlegt die SSE-Frames — reine
 * Funktionen, netzlos testbar. Form-Vorbild:
 * `examples/sse-client/stream.go`, `examples/csharp/sse-client/SseStream.cs`.
 * **Keine neue Abhängigkeit** (`ADR-0090` Festlegung 1): das Zerlegen eines
 * Frames ist Zeilen-Verarbeitung über die JDK-Standardbibliothek, kein
 * Fremdmodul (etwa `okhttp-sse`) nötig.
 */
object SseStream {
    fun streamUrl(addr: String): String = "http://$addr/changes/stream"

    /**
     * readEvent liest ein vollständiges Frame über [next] und liefert es
     * zurück. Ein Frame endet mit der Leerzeile, die der Server nach der
     * `event:`- und der `data:`-Zeile schreibt (`ADR-0061`); sie trennt zwei
     * aufeinanderfolgende Events. Ist die Quelle vor dem Frame-Abschluss
     * erschöpft ([next] liefert `null`), liefert readEvent `null`, und ein
     * begonnenes Frame wird verworfen — ein unvollständiges Frame ist kein
     * Event.
     */
    fun readEvent(next: () -> String?): Event? {
        var name: String? = null
        var data: String? = null
        while (true) {
            val line = next() ?: return null
            if (line.isEmpty()) {
                if (name == null && data == null) {
                    continue
                }
                return Event(name ?: "", data ?: "")
            }
            when {
                line.startsWith("event: ") -> name = line.removePrefix("event: ")
                line.startsWith("data: ") -> data = line.removePrefix("data: ")
            }
        }
    }
}

File: examples/csharp/sse-client/SseStream.cs
namespace CdcExamples.Sse;

/// <summary>
/// Event trägt ein Frame des Change-Streams (<c>LH-FA-SST-008</c>): den Namen
/// aus der <c>event:</c>-Zeile und die Nutzlast aus der <c>data:</c>-Zeile.
/// Der Server schreibt je Change ein JSON-Objekt mit den zehn
/// Nachrichtenfeldern in die <c>data:</c>-Zeile (<c>ADR-0061</c>).
/// </summary>
public sealed record Event(string Name, string Data);

/// <summary>
/// SseStream baut die Stream-Adresse und zerlegt die SSE-Frames — reine
/// Funktionen, netzlos testbar. Form-Vorbild: <c>examples/sse-client/stream.go</c>.
/// **Keine neue Abhängigkeit** (<c>ADR-0090</c> Festlegung 1): das Zerlegen
/// eines Frames ist Zeilen-Verarbeitung über die Runtime-Standardbibliothek,
/// kein Fremdmodul (etwa <c>System.Net.ServerSentEvents</c>) nötig.
/// </summary>
public static class SseStream
{
    /// <summary>
    /// StreamUrl baut die Adresse des SSE-Endpunkts (<c>LH-FA-SST-008</c>): ein
    /// <c>GET</c> auf <c>/changes/stream</c>. <paramref name="addr"/> ist die
    /// Horch-Adresse des Feed-Containers (<c>CDC_HTTP_ADDR</c>, Form
    /// <c>host:port</c>).
    /// </summary>
    public static string StreamUrl(string addr) => $"http://{addr}/changes/stream";

    /// <summary>
    /// ReadEvent liest ein vollständiges Frame über <paramref name="next"/> und
    /// liefert es zurück. Ein Frame endet mit der Leerzeile, die der Server
    /// nach der <c>event:</c>- und der <c>data:</c>-Zeile schreibt
    /// (<c>ADR-0061</c>); sie trennt zwei aufeinanderfolgende Events. Ist die
    /// Quelle vor dem Frame-Abschluss erschöpft (<paramref name="next"/>
    /// liefert <c>null</c>), liefert ReadEvent <c>null</c>, und ein begonnenes
    /// Frame wird verworfen — ein unvollständiges Frame ist kein Event.
    /// </summary>
    public static Event? ReadEvent(Func<string?> next)
    {
        string? name = null;
        string? data = null;
        while (true)
        {
            var line = next();
            if (line is null)
            {
                return null;
            }
            if (line.Length == 0)
            {
                if (name is null && data is null)
                {
                    continue;
                }
                return new Event(name ?? "", data ?? "");
            }
            if (line.StartsWith("event: ", StringComparison.Ordinal))
            {
                name = line["event: ".Length..];
            }
            else if (line.StartsWith("data: ", StringComparison.Ordinal))
            {
                data = line["data: ".Length..];
            }
        }
    }
}
